### PR TITLE
[Chat] Refactoring mention feature 1

### PR DESCRIPTION
**Pull request by Leah-Xia-Microsoft**
# What
Part 1 of the refactoring ticket for mention feature

Separate some function into smaller chunks
Address issues from previous PR:
https://github.com/Azure/communication-ui-library/pull/2982#discussion_r1186382630
https://github.com/Azure/communication-ui-library/pull/2982#discussion_r1186386618
https://github.com/Azure/communication-ui-library/pull/2982#discussion_r1186442301
https://github.com/Azure/communication-ui-library/pull/2982#discussion_r1186447075
https://github.com/Azure/communication-ui-library/pull/2982#discussion_r1186444434

Adjust use of null to undefined where at all possible.
https://github.com/Azure/communication-ui-library/pull/2982#discussion_r1186429934

User of forEach() instead of for i...
https://github.com/Azure/communication-ui-library/pull/2982#discussion_r1186439960

Review logic
https://github.com/Azure/communication-ui-library/pull/2982#discussion_r1186441062

Use of conditional compilation:
https://github.com/Azure/communication-ui-library/pull/2982#discussion_r1186431425
https://github.com/Azure/communication-ui-library/pull/2982#discussion_r1186431647

# Process & policy checklist

- [x] I have updated the project documentation to reflect my changes if necessary.
- [x] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: change-beta/@azure-communication-react-ad3f66ea-b008-4aaa-a540-a1b7fdef651d.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
```
Comment (vhuseinova-msft) [inline change-beta/@azure-communication-react-ad3f66ea-b008-4aaa-a540-a1b7fdef651d.json:2]: The type should be 'none'

File: packages/react-components/src/components/InputBoxComponent.tsx [hunk 2]
```diff
@@ -589,45 +445,70 @@ export const InputBoxComponent = (props: InputBoxComponentProps): JSX.Element =>
...
-          setCaretIndex(updatedContent.updatedSelectionIndex);
-          setSelectionEndValue(updatedContent.updatedSelectionIndex);
-          setSelectionStartValue(updatedContent.updatedSelectionIndex);
-        }
+    },
+    [debouncedQueryUpdate, mentionLookupOptions, updateMentionSuggestions]
+  );
+
+  /* @conditional-compile-remove(mention) */
+  const handleOnChange = useCallback(
```
Comment (vhuseinova-msft) [inline packages/react-components/src/components/InputBoxComponent.tsx:453]: As we've discussed, the object instead of just a parameters list can be a better option here

File: packages/react-components/src/components/InputBoxComponent.tsx [hunk 3]
```diff
@@ -1047,7 +928,7 @@ type MentionTagUpdateProps = {
   processedChange: string;
   change: string;
   tag: TagData;
-  closeTagIdx: number;
+  closeTagIndex: number;
```
Comment (vhuseinova-msft) [inline packages/react-components/src/components/InputBoxComponent.tsx:931]: minor: Should it be `closingTagIndex`?

File: packages/react-components/src/components/utils/keyboardNavigation.ts [hunk 4]
```diff
@@ -14,3 +14,17 @@ export const submitWithKeyboard = (
...
 };
+
+/**
+ * Determine if the press of the enter key is from a composition session or not (Safari only)
+ *
+ * @private
+ */
+export const isEnterKeyEventFromCompositionSession = (e: React.KeyboardEvent<HTMLElement>): boolean => {
+  // Uses KeyCode 229 and which code 229 to determine if the press of the enter key is from a composition session or not (Safari only)
+  if (e.nativeEvent.isComposing || e.nativeEvent.keyCode === 229 || e.nativeEvent.which === 229) {
```
Comment (vhuseinova-msft) [inline packages/react-components/src/components/utils/keyboardNavigation.ts:25]: I think the function body can be changed to
`return e.nativeEvent.isComposing || e.nativeEvent.keyCode === 229 || e.nativeEvent.which === 229`

File: packages/react-components/src/components/InputBoxComponent.tsx [hunk 5]
```diff
@@ -39,10 +39,13 @@ import { useTheme } from '../theming';
 import { MentionLookupOptions, _MentionPopover, Mention } from './MentionPopover';
 /* @conditional-compile-remove(mention) */
 import { useDebouncedCallback } from 'use-debounce';
+import { isEnterKeyEventFromCompositionSession } from './utils/keyboardNavigation';
```
Comment (vhuseinova-msft) [inline packages/react-components/src/components/InputBoxComponent.tsx:42]: Could you clarify if there will be a separate PR for moving out the mentions functionality out of `InputBoxComponent`?

**Comment by emlynmac** [comment]
Some of the checks are failing for updating snapshots. 
This PR shouldn't require updates - if there are breaks it is likely due to a change in focus or key handling and needs to be fixed up,.

File: packages/react-components/src/components/InputBoxComponent.tsx [hunk 6]
```diff
@@ -589,45 +445,70 @@ export const InputBoxComponent = (props: InputBoxComponentProps): JSX.Element =>
...
+      currentSelectionStart?: number,
+      currentSelectionEnd?: number,
+      updatedValue?: string
+    ): Promise<void> => {
+      debouncedQueryUpdate.cancel();
+      if (event.currentTarget === null) {
+        return;
       }
+      // handle backspace change
+      // onSelect is not called for backspace as selection is not changed and local caret index is outdated
```
Comment (vhuseinova-msft) [inline packages/react-components/src/components/InputBoxComponent.tsx:471]: On change functionality is broken. Backspace in the middle of the mention doesn't work anymore.
Steps:

1. Input a sentence with a couple of mentions
2. Navigate to the second mention
3. Set coursor ( by keyboard) to be in the middle of the mention
4. Backspace
5. Start to type something

Result: the coursor is moved to the end of the sentence or the mention is removed
Expected result: coursor should be set before the mention and a user can input new information if needed

File: packages/react-components/src/components/InputBoxComponent.tsx [hunk 7]
```diff
@@ -589,45 +445,70 @@ export const InputBoxComponent = (props: InputBoxComponentProps): JSX.Element =>
...
+      setCaretIndex(undefined);
```
Comment (vhuseinova-msft) [inline packages/react-components/src/components/InputBoxComponent.tsx:472]: On change functionality is broken. Delete in the beginning of the mention doesn't work anymore.
Steps:

1. Input a sentence with a mention (example: 'Hi @patriciaAdams and')
2. Navigate to the beginning of the mention
4. Delete
5. Start to type something ( at least a couple of symbols)

Result: the mention is removed
Expected result: coursor should be set before the mention and a user can input new information if needed

File: packages/react-components/src/components/InputBoxComponent.tsx [hunk 8]
```diff
@@ -589,45 +445,70 @@ export const InputBoxComponent = (props: InputBoxComponentProps): JSX.Element =>
...
+      const newValue = updatedValue ?? '';
```
Comment (vhuseinova-msft) [inline packages/react-components/src/components/InputBoxComponent.tsx:473]: Sometimes after deleting the whole string from the send box and starting to input "Hi @patricia Adams and" lead that the last symbol is inputted instead of the real keys that are selected on the keyboard. So instead of "Hi @patricia Adams and", I see "Hi @patricia Adams aaa"

File: packages/react-components/src/components/InputBoxComponent.tsx [hunk 9]
```diff
@@ -450,108 +351,63 @@ export const InputBoxComponent = (props: InputBoxComponentProps): JSX.Element =>
...
       // (because of setSelectionRange calls or rerender)
     },
-    [
-      updateSelectionIndexesWithMentionIfNeeded,
-      targetSelection,
-      setTargetSelection,
-      setSelectionStartValue,
-      setSelectionEndValue
-    ]
+    [caretIndex]
```
Comment (emlynmac) [inline packages/react-components/src/components/InputBoxComponent.tsx:399]: `setSelectionStartValue` and `setSelectionEndValue` are called above, so why have they been removed from the array here?

**Review by emlynmac — COMMENTED** [review]
Looks like some functionality needs fixing up.
